### PR TITLE
MOD-7319: Support debian:bookworm

### PR DESCRIPTION
**Pull request by nafraf**
Add support for debian:bookworm for amd64 and arm64

A clear and concise description of what the PR is solving, including:
1. Current State: Github workflows do not support debian:bookworm
2. Change: Add support for debian:bookworm to the workflows
3. Outcome: Debian:bookworm should be added to our github workflow

**Which issues this PR fixes**
1. MOD-7319

**Main objects this PR modified**
1. Github workflows

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: .install/debian_gnu_linux_12.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+if [[ $ARCH = 'x86_64' ]]
+then
+        $MODE apt install -yqq gcc-11 g++-11
+        # Update alternatives
+        $MODE update-alternatives \
+                --install /usr/bin/gcc gcc /usr/bin/gcc-11 60 \
+                --slave /usr/bin/g++ g++ /usr/bin/g++-11
+fi
+
+source install_cmake.sh $MODE
```
Comment (raz-mon) [inline .install/debian_gnu_linux_12.sh:22]: End file with new line.

File: .github/workflows/flow-temp.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ name: temporary testing
 
 on:
   push:
-    branches-ignore: ['**'] # ignore all branches. Comment this line to run your workflow below on every push.
+    # branches-ignore: ['**'] # ignore all branches. Comment this line to run your workflow below on every push.
```
Comment (raz-mon) [inline .github/workflows/flow-temp.yml:12]: Comment in

File: .github/workflows/flow-temp.yml [hunk 3]
```diff
@@ -30,18 +30,18 @@ jobs:
...
+  test-debian:
+    needs: [docs-only, get-latest-redis-tag]
+    if: needs.docs-only.outputs.only-docs-changed == 'false'
+    uses: ./.github/workflows/flow-linux-platforms.yml
     secrets: inherit
     with:
-      coordinator: true
-      test-config: QUICK=1
+      redis-ref: ${{ needs.get-latest-redis-tag.outputs.tag }}
+      platform: "debian:bookworm"
```
Comment (raz-mon) [inline .github/workflows/flow-temp.yml:40]: Comment out once you see all is good

File: .github/workflows/task-get-linux-configurations.yml [hunk 4]
```diff
@@ -9,11 +9,13 @@ env:
...
                     'rockylinux:9',
                     'debian:bullseye',
+                    'debian:bookworm',
                     'amazonlinux:2',
                     'mcr.microsoft.com/cbl-mariner/base/core:2.0']"
   ALL_ARM_IMAGES: "['ubuntu:jammy',
                     'ubuntu:focal',
-                    'ubuntu:bionic']"
+                    'ubuntu:bionic',
+                    'debian:bookworm']"
```
Comment (raz-mon) [inline .github/workflows/task-get-linux-configurations.yml:18]: Do we need arm for `bookworm`?
Sure we don't need `bullseye` here as well?

Comment (nafraf) [reply .github/workflows/task-get-linux-configurations.yml:18]: `bullseye` changes are not listed in the epic MOD-7304, but it is part of the PRD.
Additionally, we don't have json package for bullseye on arm64, then maybe we should create a new PR to support that OS/architecture.

File: .install/debian_gnu_linux_12.sh (hunk 1 above)
Comment (nafraf) [reply .install/debian_gnu_linux_12.sh]: OK. Done.

File: sbin/get-redisjson [hunk 5]
```diff
@@ -72,7 +72,9 @@ else
 			nick="ubuntu18.04"
 		fi
 	elif [[ $dist == debian ]]; then
-		if [[ $nick != bullseye ]]; then
+		if [[ $nick == bookworm ]]; then
+			nick="ubuntu22.04"
```
[the remark this answers is not in the document]
Comment (nafraf) [reply sbin/get-redisjson:76]: Change done. We only support debian `bullseye` and `bookworm`, and for `bullseye/x86_64` we have the corresponding redisjson module, then the condition is needed only for `bookworm`

File: .github/workflows/flow-temp.yml (hunk 2 above)
Comment (nafraf) [reply .github/workflows/flow-temp.yml]: OK. Changes in flow-temp.yml were reverted.

File: .github/workflows/flow-temp.yml (hunk 3 above)
Comment (nafraf) [reply .github/workflows/flow-temp.yml]: OK. Changes in flow-temp.yml were reverted.

File: .install/debian_gnu_linux_12.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+# Install gcc-11 and g++-11 because gcc-12.2 fails to compile VectorSimilarity
+# https://gcc.gnu.org/bugzilla/show_bug.cgi?id=105593
+if [[ $ARCH = 'x86_64' ]]
+then
+        $MODE apt install -yqq gcc-11 g++-11
+        # Update alternatives
+        $MODE update-alternatives \
+                --install /usr/bin/gcc gcc /usr/bin/gcc-11 60 \
+                --slave /usr/bin/g++ g++ /usr/bin/g++-11
+fi
```
Comment (GuyAv46) [inline .install/debian_gnu_linux_12.sh:20]: can we address Meirav's message and use GCC 12.3/13? attempting to update GCC with `apt install -yqq gcc` or `apt upgrade gcc` doesn't fetch 12.3?

Comment (nafraf) [reply .install/debian_gnu_linux_12.sh:20]: apt does not fetch 12.3, I tried with these repositories, but none of them contains 12.3:
[bookworm](https://packages.debian.org/bookworm/)
[bookworm-updates](https://packages.debian.org/bookworm-updates/)
[bookworm-backports](https://packages.debian.org/bookworm-backports/)
Should I use a different repository?

Comment (raz-mon) [reply .install/debian_gnu_linux_12.sh:20]: Could [this](https://packages.debian.org/trixie/amd64/gcc/download) be relevant?

Comment (raz-mon) [reply .install/debian_gnu_linux_12.sh:20]: Possibly [this](https://hub.docker.com/layers/library/gcc/12.4.0-bookworm/images/sha256-1f21186263c2a5823c0ee81033d2980256a958af8a53801e7baf5744e07ff80f?context=explore) as well

Comment (GuyAv46) [reply .install/debian_gnu_linux_12.sh:20]: We can also try setting a make/cmake arg that will silent the warning (specifically for this OS on CI), or add back the ignore flag to vecsim cmake

File: .github/workflows/task-get-linux-configurations.yml [hunk 7]
```diff
@@ -64,7 +64,10 @@ jobs:
           def needs_arch(arch):
             return requested_arch in ['all', arch]
 
-          naming = {'mariner:2': 'mcr.microsoft.com/cbl-mariner/base/core:2.0'}
+          naming = {
+            'mariner:2': 'mcr.microsoft.com/cbl-mariner/base/core:2.0',
+            'debian:bookworm': 'gcc:12.4.0-bookworm'
```
Comment (GuyAv46) [inline .github/workflows/task-get-linux-configurations.yml:69]: I think we should track `gcc:12.4-bookworm` or `gcc:12-bookworm`

File: .github/workflows/task-get-linux-configurations.yml [hunk 8]
```diff
@@ -9,7 +9,7 @@ env:
                     'ubuntu:bionic',
                     'rockylinux:8',
                     'rockylinux:9',
-                    'debian:bullseye',
+                    'bullseye',
```
Comment (GuyAv46) [inline .github/workflows/task-get-linux-configurations.yml:12]: Why is this change needed?

File: .github/workflows/task-get-linux-configurations.yml (hunk 7 above)
Comment (raz-mon) [reply .github/workflows/task-get-linux-configurations.yml:69]: Agreed, `12.4` in my opinion.

File: .install/debian_gnu_linux_12.sh (hunk 6 above)
Comment (raz-mon) [inline .install/debian_gnu_linux_12.sh:20]: Removing this?

Comment (nafraf) [reply .install/debian_gnu_linux_12.sh]: Yes. Done. Thanks.

**Review by GuyAv46 — APPROVED** [review]
Nice

**Review by raz-mon — APPROVED** [review]
